### PR TITLE
Enforce Checksum on Sign

### LINKED ISSUE #712: Enforce checksum address on sign
- I think you can just do it here https://github.com/snapshot-labs/snapshot.js/blob/master/src/sign/index.ts#L55
- using https://docs.ethers.io/v5/api/utils/address/#utils-getAddress from https://www.npmjs.com/package/@ethersproject/address
- We have to check all methods in this class (vote/proposal/settings etc. ) should make sure they work fine even if we pass an address in any case
- To run these methods you can follow https://docs.snapshot.org/snapshot.js

### PR DESCRIPTION
**Pull request by emanuel-sol**
Fixes #712 .

Changes proposed in this pull request:
- Add ethers/address submodule in order to use getAddress function
- Preprocess address input by lowercasing it to prevent runtime errors

I was able to successfully test the changes in the snapshot UI locally using `npm link`, but I'm having a difficult time mocking requests with malformed case addresses. I could use some testing guidance here.

### CODE REVIEW

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/sign/index.ts [hunk 1]
```diff
@@ -52,13 +53,14 @@ export default class Client {
...
+    const checksumAddress = getAddress(address.toLowerCase());
+    if (!message.from) message.from = checksumAddress;
     if (!message.timestamp)
       message.timestamp = parseInt((Date.now() / 1e3).toFixed());
     const data: any = { domain, types, message };
     const sig = await signer._signTypedData(domain, data.types, message);
-    // console.log('Sign', { address, sig, data });
-    return await this.send({ address, sig, data });
+    // console.log('Sign', { checksumAddress, sig, data });
+    return await this.send({ checksumAddress, sig, data });
```
[the remark this answers is not in the document]
Comment (emanuel-sol) [reply src/sign/index.ts]: Ah yes, I forgot ES6 object destructuring would change the property name. Snapshot-Hub would reject the promise as defined [here](https://github.com/snapshot-labs/snapshot-hub/blob/e810bffb72b20db38fd4bb727f5a3773c49df4f6/src/ingestor/typedData/index.ts#L18). I hard coded the property name to be `address` as a fix.

File: src/sign/index.ts [hunk 2]
```diff
@@ -52,13 +53,14 @@ export default class Client {
   async sign(web3: Web3Provider | Wallet, address: string, message, types) {
     // @ts-ignore
     const signer = web3?.getSigner ? web3.getSigner() : web3;
-    if (!message.from) message.from = address;
+    const checksumAddress = getAddress(address.toLowerCase());
+    if (!message.from) message.from = checksumAddress;
```
Comment (ChaituVR) [inline src/sign/index.ts:57]: I feel we should do checksum for `message.from` too.

File: src/sign/index.ts [hunk 3]
```diff
@@ -52,13 +53,14 @@ export default class Client {
   async sign(web3: Web3Provider | Wallet, address: string, message, types) {
     // @ts-ignore
     const signer = web3?.getSigner ? web3.getSigner() : web3;
-    if (!message.from) message.from = address;
+    const checksumAddress = getAddress(address.toLowerCase());
```
Comment (bonustrack) [inline src/sign/index.ts:56]: Why toLowercase here?

File: src/sign/index.ts (hunk 2 above)
Comment (bonustrack) [reply src/sign/index.ts:57]: Of course

Comment (bonustrack) [reply src/sign/index.ts:57]: This checksum message.from only if its not part of the message, if its part of the message it wont be checksumed you should change that

File: src/sign/index.ts (hunk 3 above)
Comment (emanuel-sol) [reply src/sign/index.ts:56]: I used the lowercase function for basic error checking in case we receive a malformed checksum address

Comment (bonustrack) [reply src/sign/index.ts:56]: In which case does it throw and error here? I dont think toLowercase change anything here

File: src/sign/index.ts [hunk 4]
```diff
@@ -52,13 +53,18 @@ export default class Client {
...
+    if (message.from) {
```
Comment (bonustrack) [inline src/sign/index.ts:57]: This could be made much simpler

File: src/sign/index.ts (hunk 3 above)
Comment (emanuel-sol) [reply src/sign/index.ts:56]: Yeah it's probably a good idea if we throw an error if we get a bad checksum address. I'll get rid of it

**Review by ChaituVR — APPROVED** [review]
Tested ACK

**Pull request merged**